# 参考资料
[DataWhale组队学习](http://datawhale.club/t/topic/580)

# EX1

In [285]:
ex1 = pd.read_csv('../data/company.csv')

In [286]:
ex1.head()

EmployeeID birthdate_key  age  city_name department           job_title  \
0        1318      1/3/1954   61  Vancouver  Executive                 CEO   
1        1319      1/3/1957   58  Vancouver  Executive           VP Stores   
2        1320      1/2/1955   60  Vancouver  Executive       Legal Counsel   
3        1321      1/2/1959   56  Vancouver  Executive  VP Human Resources   
4        1322      1/9/1958   57  Vancouver  Executive          VP Finance   

  gender  
0      M  
1      F  
2      F  
3      M  
4      M

### 1）分别只使用 query 和 loc 选出年龄不超过四十岁且工作部门为 Dairy 或 Bakery 的男性。

In [287]:
adt1 = ex1.age <= 40
adt2 = ex1.department.isin(["Dairy", "Bakery"])
adt3 = ex1.gender == "M"
ex1.loc[adt1 & adt2 & adt3].head()

EmployeeID birthdate_key  age city_name department     job_title gender
3611        5791     1/14/1975   40   Kelowna      Dairy  Dairy Person      M
3613        5793     1/22/1975   40  Richmond     Bakery         Baker      M
3615        5795     1/30/1975   40   Nanaimo      Dairy  Dairy Person      M
3617        5797      2/3/1975   40   Nanaimo      Dairy  Dairy Person      M
3618        5798      2/4/1975   40    Surrey      Dairy  Dairy Person      M

In [288]:
ex1.query("(age <= 40) & (department == ['Dairy', 'Bakery']) & (gender == 'M')").head()

EmployeeID birthdate_key  age city_name department     job_title gender
3611        5791     1/14/1975   40   Kelowna      Dairy  Dairy Person      M
3613        5793     1/22/1975   40  Richmond     Bakery         Baker      M
3615        5795     1/30/1975   40   Nanaimo      Dairy  Dairy Person      M
3617        5797      2/3/1975   40   Nanaimo      Dairy  Dairy Person      M
3618        5798      2/4/1975   40    Surrey      Dairy  Dairy Person      M

### 2) 选出员工 ID 号为奇数所在行的第 1、第 3 和倒数第 2 列

In [289]:
# 模数为0是偶数
ex1.EmployeeID % 2 == 1

0       False
1        True
2       False
3        True
4       False
        ...  
6279    False
6280     True
6281     True
6282    False
6283    False
Name: EmployeeID, Length: 6284, dtype: bool

In [290]:
x = ex1.loc[ex1.EmployeeID % 2 == 1]
x.iloc[:, [0,2,-2]]

EmployeeID  age                      job_title
1           1319   58                      VP Stores
3           1321   56             VP Human Resources
5           1323   53      Exec Assistant, VP Stores
6           1325   51  Exec Assistant, Legal Counsel
8           1329   48                  Store Manager
...          ...  ...                            ...
6276        7659   26                        Cashier
6277        7741   25                        Cashier
6278        7801   25                   Dairy Person
6280        8181   22                        Cashier
6281        8223   21                        Cashier

[3126 rows x 3 columns]

In [291]:
ex1.iloc[(ex1.EmployeeID % 2 == 1).values, [0, 2,-2]]

EmployeeID  age                      job_title
1           1319   58                      VP Stores
3           1321   56             VP Human Resources
5           1323   53      Exec Assistant, VP Stores
6           1325   51  Exec Assistant, Legal Counsel
8           1329   48                  Store Manager
...          ...  ...                            ...
6276        7659   26                        Cashier
6277        7741   25                        Cashier
6278        7801   25                   Dairy Person
6280        8181   22                        Cashier
6281        8223   21                        Cashier

[3126 rows x 3 columns]

### 3)  按照以下步骤进行索引操作：
```
• 把后三列设为索引后交换内外两层
• 恢复中间一层
• 修改外层索引名为 Gender
• 用下划线合并两层行索引
• 把行索引拆分为原状态
• 修改索引名为原表名称
• 恢复默认索引并将列保持为原表的相对位置
```

In [292]:
ex1_ = ex1.copy()

In [293]:
x = ex1_.columns
print(x, x[-3 : -1], x[-3 :], sep = "\n\n")

Index(['EmployeeID', 'birthdate_key', 'age', 'city_name', 'department',
       'job_title', 'gender'],
      dtype='object')

Index(['department', 'job_title'], dtype='object')

Index(['department', 'job_title', 'gender'], dtype='object')


In [294]:
ex1_.set_index(ex1_.columns[-3:].tolist())

EmployeeID birthdate_key  age  \
department       job_title          gender                                  
Executive        CEO                M             1318      1/3/1954   61   
                 VP Stores          F             1319      1/3/1957   58   
                 Legal Counsel      F             1320      1/2/1955   60   
                 VP Human Resources M             1321      1/2/1959   56   
                 VP Finance         M             1322      1/9/1958   57   
...                                                ...           ...  ...   
Customer Service Cashier            F             8036      8/9/1992   23   
                                    M             8181     9/26/1993   22   
                                    M             8223     2/11/1994   21   
                                    F             8226     2/16/1994   21   
                                    F             8264     6/13/1994   21   

                                                   city_name  
department       job_title          gender                    
Executive        CEO                M              Vancouver  
                 VP Stores          F              Vancouver  
                 Legal Counsel      F              Vancouver  
                 VP Human Resources M              Vancouver  
                 VP Finance         M              Vancouver  
...                                                      ...  
Customer Service Cashier            F       New Westminister  
                                    M          Prince George  
                                    M                  Trail  
                                    F               Victoria  
                                    F              Vancouver  

[6284 rows x 4 columns]

In [295]:
ex1_ = ex1_.set_index(ex1_.columns[-3:].tolist()).swaplevel(0,2,axis=0)
ex1_.head()

EmployeeID birthdate_key  age  city_name
gender job_title          department                                          
M      CEO                Executive         1318      1/3/1954   61  Vancouver
F      VP Stores          Executive         1319      1/3/1957   58  Vancouver
       Legal Counsel      Executive         1320      1/2/1955   60  Vancouver
M      VP Human Resources Executive         1321      1/2/1959   56  Vancouver
       VP Finance         Executive         1322      1/9/1958   57  Vancouver

In [296]:
ex1_ = ex1_.reset_index(level=1)
ex1_.head()

job_title  EmployeeID birthdate_key  age  \
gender department                                                      
M      Executive                  CEO        1318      1/3/1954   61   
F      Executive            VP Stores        1319      1/3/1957   58   
       Executive        Legal Counsel        1320      1/2/1955   60   
M      Executive   VP Human Resources        1321      1/2/1959   56   
       Executive           VP Finance        1322      1/9/1958   57   

                   city_name  
gender department             
M      Executive   Vancouver  
F      Executive   Vancouver  
       Executive   Vancouver  
M      Executive   Vancouver  
       Executive   Vancouver

In [297]:
ex1_ = ex1_.rename_axis(index={'gender':'Gender'})
ex1_.head()

job_title  EmployeeID birthdate_key  age  \
Gender department                                                      
M      Executive                  CEO        1318      1/3/1954   61   
F      Executive            VP Stores        1319      1/3/1957   58   
       Executive        Legal Counsel        1320      1/2/1955   60   
M      Executive   VP Human Resources        1321      1/2/1959   56   
       Executive           VP Finance        1322      1/9/1958   57   

                   city_name  
Gender department             
M      Executive   Vancouver  
F      Executive   Vancouver  
       Executive   Vancouver  
M      Executive   Vancouver  
       Executive   Vancouver

In [298]:
ex1_.index = ex1_.index.map(lambda x:'_'.join(x))
ex1_.head()

job_title  EmployeeID birthdate_key  age  city_name
M_Executive                 CEO        1318      1/3/1954   61  Vancouver
F_Executive           VP Stores        1319      1/3/1957   58  Vancouver
F_Executive       Legal Counsel        1320      1/2/1955   60  Vancouver
M_Executive  VP Human Resources        1321      1/2/1959   56  Vancouver
M_Executive          VP Finance        1322      1/9/1958   57  Vancouver

In [299]:
ex1_index = ex1_.index.map(lambda x:tuple(x.split('_')))
ex1_.head()

job_title  EmployeeID birthdate_key  age  city_name
M_Executive                 CEO        1318      1/3/1954   61  Vancouver
F_Executive           VP Stores        1319      1/3/1957   58  Vancouver
F_Executive       Legal Counsel        1320      1/2/1955   60  Vancouver
M_Executive  VP Human Resources        1321      1/2/1959   56  Vancouver
M_Executive          VP Finance        1322      1/9/1958   57  Vancouver

# EX2

In [321]:
ex2 = pd.read_csv('../data/chocolate.csv')

In [322]:
ex2.head()

Company  Review\nDate Cocoa\nPercent Company\nLocation  Rating
0  A. Morin          2016            63%            France    3.75
1  A. Morin          2015            70%            France    2.75
2  A. Morin          2015            70%            France    3.00
3  A. Morin          2015            70%            France    3.50
4  A. Morin          2015            70%            France    3.50

### 1. 把列索引名中的 \n 替换为空格。

In [323]:
ex2.columns = [' '.join(i.split('\n')) for i in ex2.columns]
ex2.head()

Company  Review Date Cocoa Percent Company Location  Rating
0  A. Morin         2016           63%           France    3.75
1  A. Morin         2015           70%           France    2.75
2  A. Morin         2015           70%           France    3.00
3  A. Morin         2015           70%           France    3.50
4  A. Morin         2015           70%           France    3.50

[join函数](https://www.runoob.com/python/att-string-join.html)

In [319]:
str = "-";
seq = ["a", "b", "c"]; # 字符串序列
print (str.join( seq ))

a-b-c


### 2. 巧克力 Rating 评分为 1 至 5，每 0.25 分一档，请选出 2.75 分及以下且可可含量 Cocoa Percent 高于中位数的样本。

### 3. 将 Review Date 和 Company Location 设为索引后，选出 Review Date 在 2012 年之后且 Company
Location 不属于 France, Canada, Amsterdam, Belgium 的样本。

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('../data/learn_pandas.csv'
                  , usecols = ['School', 'Grade', 'Name', 'Gender', 'Weight', 'Transfer'])


In [3]:
df.head()

School      Grade            Name  Gender  Weight  \
0  Shanghai Jiao Tong University   Freshman    Gaopeng Yang  Female    46.0   
1              Peking University   Freshman  Changqiang You    Male    70.0   
2  Shanghai Jiao Tong University     Senior         Mei Sun    Male    89.0   
3               Fudan University  Sophomore    Xiaojuan Sun  Female    41.0   
4               Fudan University  Sophomore     Gaojuan You    Male    74.0   

  Transfer  
0        N  
1        N  
2        N  
3        N  
4        N

In [4]:
df["Name"].head()

0      Gaopeng Yang
1    Changqiang You
2           Mei Sun
3      Xiaojuan Sun
4       Gaojuan You
Name: Name, dtype: object

In [5]:
df[["Name", "School"]].head()

Name                         School
0    Gaopeng Yang  Shanghai Jiao Tong University
1  Changqiang You              Peking University
2         Mei Sun  Shanghai Jiao Tong University
3    Xiaojuan Sun               Fudan University
4     Gaojuan You               Fudan University

In [6]:
df.Name.head()

0      Gaopeng Yang
1    Changqiang You
2           Mei Sun
3      Xiaojuan Sun
4       Gaojuan You
Name: Name, dtype: object

In [7]:
s = pd.Series(data = [i for i in range(5)]
             , index = ["b", "a", "c", "a", "a"])
print(s, s["a"], s["b"], s[["b", "c"]], sep = "\n\n")

b    0
a    1
c    2
a    3
a    4
dtype: int64

a    1
a    3
a    4
dtype: int64

0

b    0
c    2
dtype: int64


In [8]:
# 注意两个索引是单索引， 否则出现错误  如： s["a" : "c" : 1]
# 且右端取出
s["b" : "c" : 1]

b    0
a    1
c    2
dtype: int64

In [9]:
s = pd.Series(['a', 'b', 'c', 'd', 'e', 'f']
                , index=[1, 3, 1, 2, 5, 4])
# 注意当索引为整数的时候， 切片取出的是对应的下标， 且右端不取出
print(s, s[1], s[[2, 1]], s[1 : -1 : 2], sep = "\n\n")

1    a
3    b
1    c
2    d
5    e
4    f
dtype: object

1    a
1    c
dtype: object

2    d
1    a
1    c
dtype: object

3    b
2    d
dtype: object


In [10]:
df.head()

School      Grade            Name  Gender  Weight  \
0  Shanghai Jiao Tong University   Freshman    Gaopeng Yang  Female    46.0   
1              Peking University   Freshman  Changqiang You    Male    70.0   
2  Shanghai Jiao Tong University     Senior         Mei Sun    Male    89.0   
3               Fudan University  Sophomore    Xiaojuan Sun  Female    41.0   
4               Fudan University  Sophomore     Gaojuan You    Male    74.0   

  Transfer  
0        N  
1        N  
2        N  
3        N  
4        N

In [11]:
# 将姓名列设置为索引名
df = df.set_index("Name")
df.head()

School      Grade  Gender  Weight  \
Name                                                                       
Gaopeng Yang    Shanghai Jiao Tong University   Freshman  Female    46.0   
Changqiang You              Peking University   Freshman    Male    70.0   
Mei Sun         Shanghai Jiao Tong University     Senior    Male    89.0   
Xiaojuan Sun                 Fudan University  Sophomore  Female    41.0   
Gaojuan You                  Fudan University  Sophomore    Male    74.0   

               Transfer  
Name                     
Gaopeng Yang          N  
Changqiang You        N  
Mei Sun               N  
Xiaojuan Sun          N  
Gaojuan You           N

In [12]:
# 返回Series
df.loc["Gaopeng Yang"]

School      Shanghai Jiao Tong University
Grade                            Freshman
Gender                             Female
Weight                                 46
Transfer                                N
Name: Gaopeng Yang, dtype: object

In [13]:
# 返回DataFrame
df.loc["Mei Sun"]  

School   Grade  Gender  Weight Transfer
Name                                                                   
Mei Sun  Shanghai Jiao Tong University  Senior    Male    89.0        N
Mei Sun  Shanghai Jiao Tong University  Junior  Female    50.0        N

In [14]:
# 同时选择行和列, 返回 Series
df.loc["Mei Sun", "Weight"]

Name
Mei Sun    89.0
Mei Sun    50.0
Name: Weight, dtype: float64

In [15]:
# 返回单个元素
df.loc['Gaopeng Yang', 'School']

'Shanghai Jiao Tong University'

In [16]:
# 取出多个行和列对应的元素
df.loc[['Qiang Sun','Quan Zhao'], ['School','Gender']]

School  Gender
Name                                            
Qiang Sun            Tsinghua University  Female
Qiang Sun            Tsinghua University  Female
Qiang Sun  Shanghai Jiao Tong University  Female
Quan Zhao  Shanghai Jiao Tong University  Female

In [17]:
df.columns

Index(['School', 'Grade', 'Gender', 'Weight', 'Transfer'], dtype='object')

In [18]:
# 切片
df.loc['Gaojuan You':'Gaoqiang Qian', 'School':'Gender']

School      Grade  Gender
Name                                                           
Gaojuan You                 Fudan University  Sophomore    Male
Xiaoli Qian              Tsinghua University   Freshman  Female
Qiang Chu      Shanghai Jiao Tong University   Freshman  Female
Gaoqiang Qian            Tsinghua University     Junior  Female

In [19]:
df_ = df.copy()
df_.index = range(df_.shape[0])
df_.head()

School      Grade  Gender  Weight Transfer
0  Shanghai Jiao Tong University   Freshman  Female    46.0        N
1              Peking University   Freshman    Male    70.0        N
2  Shanghai Jiao Tong University     Senior    Male    89.0        N
3               Fudan University  Sophomore  Female    41.0        N
4               Fudan University  Sophomore    Male    74.0        N

In [20]:
# DataFrame 使用整数索引, 仍是元素索引, 注意
df_.loc[1 : 5, "Grade" : "Weight"]

Grade  Gender  Weight
1   Freshman    Male    70.0
2     Senior    Male    89.0
3  Sophomore  Female    41.0
4  Sophomore    Male    74.0
5   Freshman  Female    51.0

In [21]:
df.Weight > 70

Name
Gaopeng Yang      False
Changqiang You    False
Mei Sun            True
Xiaojuan Sun      False
Gaojuan You        True
                  ...  
Xiaojuan Sun      False
Li Zhao           False
Chengqiang Chu    False
Chengmei Shen      True
Chunpeng Lv       False
Name: Weight, Length: 200, dtype: bool

In [22]:
# 布尔索引, 传入布尔列表必须DataFrame长度一致
df[df.Weight > 70].head()

School      Grade Gender  Weight  \
Name                                                                     
Mei Sun        Shanghai Jiao Tong University     Senior   Male    89.0   
Gaojuan You                 Fudan University  Sophomore   Male    74.0   
Xiaopeng Zhou  Shanghai Jiao Tong University   Freshman   Male    74.0   
Xiaofeng Sun             Tsinghua University     Senior   Male    71.0   
Qiang Zheng    Shanghai Jiao Tong University     Senior   Male    87.0   

              Transfer  
Name                    
Mei Sun              N  
Gaojuan You          N  
Xiaopeng Zhou        N  
Xiaofeng Sun         N  
Qiang Zheng          N

### isin方法

In [23]:
# 返回的是一个布尔Series
df.Grade.isin(['Freshman', 'Senior'])

Name
Gaopeng Yang       True
Changqiang You     True
Mei Sun            True
Xiaojuan Sun      False
Gaojuan You       False
                  ...  
Xiaojuan Sun      False
Li Zhao            True
Chengqiang Chu     True
Chengmei Shen      True
Chunpeng Lv       False
Name: Grade, Length: 200, dtype: bool

In [24]:
df[df.Grade.isin(['Freshman', 'Senior'])].head()

School     Grade  Gender  Weight  \
Name                                                                      
Gaopeng Yang    Shanghai Jiao Tong University  Freshman  Female    46.0   
Changqiang You              Peking University  Freshman    Male    70.0   
Mei Sun         Shanghai Jiao Tong University    Senior    Male    89.0   
Xiaoli Qian               Tsinghua University  Freshman  Female    51.0   
Qiang Chu       Shanghai Jiao Tong University  Freshman  Female    52.0   

               Transfer  
Name                     
Gaopeng Yang          N  
Changqiang You        N  
Mei Sun               N  
Xiaoli Qian           N  
Qiang Chu             N

|（或） , &（且） , ~（取反） 的组合来实现

In [25]:
# 多种条件组合
# 选出复旦大学中体重超过 70kg的大四学生, 或者北大男生中体重超过 80kg 的非大四的学生
cdt11 = df.School == "Fudan University"
cdt12 = df.Grade == "Senior"
cdt13 = df.Weight > 70

cdt21 = df.School == "Peking University"
cdt22 = df.Grade != "Senior"
cdt23 = df.Weight > 80

df[(cdt11 & cdt12 & cdt13) | (cdt21 & cdt22 & cdt23)].head()

School     Grade Gender  Weight Transfer
Name                                                               
Qiang Han       Peking University  Freshman   Male    87.0        N
Chengpeng Zhou   Fudan University    Senior   Male    81.0        N
Changpeng Zhao  Peking University  Freshman   Male    83.0        N
Chengpeng Qian   Fudan University    Senior   Male    73.0        Y

In [26]:
df.dtypes

School       object
Grade        object
Gender       object
Weight      float64
Transfer     object
dtype: object

In [27]:
df.select_dtypes("float").head()

Weight
Name                  
Gaopeng Yang      46.0
Changqiang You    70.0
Mei Sun           89.0
Xiaojuan Sun      41.0
Gaojuan You       74.0

In [28]:
def condition(x):
    condition_1_1 = x.School == 'Fudan University'
    condition_1_2 = x.Grade == 'Senior'
    condition_1_3 = x.Weight > 70
    condition_1 = condition_1_1 & condition_1_2 & condition_1_3
    condition_2_1 = x.School == 'Peking University'
    condition_2_2 = x.Grade == 'Senior'
    condition_2_3 = x.Weight > 80
    condition_2 = condition_2_1 & (~condition_2_2) & condition_2_3
    result = condition_1 | condition_2
    return result
# df.loc[condition(df)]
df.loc[condition]

School     Grade Gender  Weight Transfer
Name                                                               
Qiang Han       Peking University  Freshman   Male    87.0        N
Chengpeng Zhou   Fudan University    Senior   Male    81.0        N
Changpeng Zhao  Peking University  Freshman   Male    83.0        N
Chengpeng Qian   Fudan University    Senior   Male    73.0        Y

In [29]:
df.loc[lambda x:'Quan Zhao', lambda x:'Gender']

'Female'

In [30]:
df.loc['Quan Zhao', 'Gender']

'Female'

[slice函数](https://www.runoob.com/python/python-func-slice.html)

In [45]:
my_qp = slice(0, 9, 2)
arr = [i for i in range(10)]
print(arr)
arr[my_qp]

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


[0, 2, 4, 6, 8]

In [48]:
df_demo = df.copy()
# slice切片
df_demo.loc[lambda x: slice('Gaojuan You', 'Gaoqiang Qian')]
# df_demo.loc['Gaojuan You' : 'Gaoqiang Qian']

School      Grade  Gender  Weight  \
Name                                                                      
Gaojuan You                 Fudan University  Sophomore    Male    74.0   
Xiaoli Qian              Tsinghua University   Freshman  Female    51.0   
Qiang Chu      Shanghai Jiao Tong University   Freshman  Female    52.0   
Gaoqiang Qian            Tsinghua University     Junior  Female    50.0   

              Transfer  
Name                    
Gaojuan You          N  
Xiaoli Qian          N  
Qiang Chu            N  
Gaoqiang Qian        N

In [55]:
# 第一行第一列
df.iloc[0, 0]

'Shanghai Jiao Tong University'

In [57]:
df_chain = pd.DataFrame([["x","yy"],["a","cc"],["y","ww"], ], columns=list('AB'))
df_chain

A   B
0  x  yy
1  a  cc
2  y  ww

In [59]:
# 前两行， 第二列
df_chain.iloc[[0, 1], 1]

0    yy
1    cc
Name: B, dtype: object

In [60]:
# 切片， 不包括右端点
df_chain.iloc[1:3, 1]

1    cc
2    ww
Name: B, dtype: object

In [66]:
df_chain[lambda x : slice(1, 3)]

A   B
1  a  cc
2  y  ww

In [67]:
df_demo.iloc[(df_demo.Weight>80).values].head()

School      Grade Gender  Weight  \
Name                                                                      
Mei Sun         Shanghai Jiao Tong University     Senior   Male    89.0   
Qiang Zheng     Shanghai Jiao Tong University     Senior   Male    87.0   
Qiang Han                   Peking University   Freshman   Male    87.0   
Chengpeng Zhou               Fudan University     Senior   Male    81.0   
Feng Han        Shanghai Jiao Tong University  Sophomore   Male    82.0   

               Transfer  
Name                     
Mei Sun               N  
Qiang Zheng           N  
Qiang Han             N  
Chengpeng Zhou        N  
Feng Han              N

In [78]:
df_demo[df_demo.Weight>80]
# df_demo.loc[df_demo.Weight>80]

School      Grade Gender  Weight  \
Name                                                                      
Mei Sun         Shanghai Jiao Tong University     Senior   Male    89.0   
Qiang Zheng     Shanghai Jiao Tong University     Senior   Male    87.0   
Qiang Han                   Peking University   Freshman   Male    87.0   
Chengpeng Zhou               Fudan University     Senior   Male    81.0   
Feng Han        Shanghai Jiao Tong University  Sophomore   Male    82.0   
Changpeng Zhao              Peking University   Freshman   Male    83.0   
Chunli Zhao     Shanghai Jiao Tong University   Freshman   Male    83.0   
Gaoli Zhao      Shanghai Jiao Tong University     Senior   Male    83.0   

               Transfer  
Name                     
Mei Sun               N  
Qiang Zheng           N  
Qiang Han             N  
Chengpeng Zhou        N  
Feng Han              N  
Changpeng Zhao        N  
Chunli Zhao           N  
Gaoli Zhao            N

In [98]:
df.query('Grade == ["Junior", "Senior"]').head()
# df[df.Grade.isin(["Junior", "Senior"])].head()

School   Grade  Gender  Weight Transfer
Name                                                                         
Mei Sun        Shanghai Jiao Tong University  Senior    Male    89.0        N
Gaoqiang Qian            Tsinghua University  Junior  Female    50.0        N
Juan Xu                    Peking University  Junior  Female     NaN        N
Xiaoquan Lv              Tsinghua University  Junior  Female    43.0        N
Peng You       Shanghai Jiao Tong University  Senior  Female    48.0      NaN

In [111]:
df.query('Grade not in ["Freshman", "Sophomore"] &'
'Gender == "Male"').head()

School   Grade Gender  Weight Transfer
Name                                                                        
Mei Sun        Shanghai Jiao Tong University  Senior   Male    89.0        N
Xiaoqiang Qin            Tsinghua University  Junior   Male    68.0        N
Peng Wang                Tsinghua University  Junior   Male    65.0        N
Xiaofeng Sun             Tsinghua University  Senior   Male    71.0        N
Xiaopeng Shen  Shanghai Jiao Tong University  Senior   Male    62.0      NaN

In [118]:
df_sample = pd.DataFrame({'id': list('abcde'), 'value': [1, 2, 3, 4, 90]})
print(df_sample)
df_sample.sample(3, replace = True, weights = df_sample.value)

  id  value
0  a      1
1  b      2
2  c      3
3  d      4
4  e     90


id  value
4  e     90
4  e     90
4  e     90

In [117]:
df_sample.sample(3, replace = False, weights = df_sample.value)

id  value
4  e     90
3  d      4
2  c      3

In [120]:
np.random.seed(0)
multi_index = pd.MultiIndex.from_product([list('ABCD')                                       
                            , df.Gender.unique()], names=('School', 'Gender'))

multi_column = pd.MultiIndex.from_product([['Height', 'Weight']
                            , df.Grade.unique()], names=('Indicator', 'Grade'))



df_multi = pd.DataFrame(np.c_[(np.random.randn(8,4)*5 + 163).tolist()
                            , (np.random.randn(8,4)*5 + 65).tolist()]
                            , index = multi_index
                            , columns = multi_column).round(1)

df_multi

Indicator       Height                           Weight                   \
Grade         Freshman Senior Sophomore Junior Freshman Senior Sophomore   
School Gender                                                              
A      Female    171.8  165.0     167.9  174.2     60.6   55.1      63.3   
       Male      172.3  158.1     167.8  162.2     71.2   71.0      63.1   
B      Female    162.5  165.1     163.7  170.3     59.8   57.9      56.5   
       Male      166.8  163.6     165.2  164.7     62.5   62.8      58.7   
C      Female    170.5  162.0     164.6  158.7     56.9   63.9      60.5   
       Male      150.2  166.3     167.3  159.3     62.4   59.1      64.9   
D      Female    174.3  155.7     163.2  162.1     65.3   66.5      61.8   
       Male      170.7  170.3     163.8  164.9     61.6   63.2      60.9   

Indicator             
Grade         Junior  
School Gender         
A      Female   65.8  
       Male     63.5  
B      Female   74.8  
       Male     68.9  
C      Female   66.9  
       Male     67.1  
D      Female   63.2  
       Male     56.4

In [123]:
print(df_multi.index, df_multi.columns, sep = "\n\n")

MultiIndex([('A', 'Female'),
            ('A',   'Male'),
            ('B', 'Female'),
            ('B',   'Male'),
            ('C', 'Female'),
            ('C',   'Male'),
            ('D', 'Female'),
            ('D',   'Male')],
           names=['School', 'Gender'])

MultiIndex([('Height',  'Freshman'),
            ('Height',    'Senior'),
            ('Height', 'Sophomore'),
            ('Height',    'Junior'),
            ('Weight',  'Freshman'),
            ('Weight',    'Senior'),
            ('Weight', 'Sophomore'),
            ('Weight',    'Junior')],
           names=['Indicator', 'Grade'])


In [127]:
print(df_multi.index.names, df_multi.columns.names, sep = "\n\n")

['School', 'Gender']

['Indicator', 'Grade']


In [131]:
print(df_multi.index.values, df_multi.columns.values, sep = "\n\n")

[('A', 'Female') ('A', 'Male') ('B', 'Female') ('B', 'Male')
 ('C', 'Female') ('C', 'Male') ('D', 'Female') ('D', 'Male')]

[('Height', 'Freshman') ('Height', 'Senior') ('Height', 'Sophomore')
 ('Height', 'Junior') ('Weight', 'Freshman') ('Weight', 'Senior')
 ('Weight', 'Sophomore') ('Weight', 'Junior')]


In [137]:
print(df_multi.columns.get_level_values(0)
      , df_multi.columns.get_level_values(1)
      , df_multi.index.get_level_values(0)
      , df_multi.index.get_level_values(1)
      , sep = "\n\n")

Index(['Height', 'Height', 'Height', 'Height', 'Weight', 'Weight', 'Weight',
       'Weight'],
      dtype='object', name='Indicator')

Index(['Freshman', 'Senior', 'Sophomore', 'Junior', 'Freshman', 'Senior',
       'Sophomore', 'Junior'],
      dtype='object', name='Grade')

Index(['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D'], dtype='object', name='School')

Index(['Female', 'Male', 'Female', 'Male', 'Female', 'Male', 'Female', 'Male'], dtype='object', name='Gender')


In [164]:
df = pd.read_csv('../data/learn_pandas.csv'
                  , usecols = ['School', 'Grade', 'Name', 'Gender', 'Weight', 'Transfer'])
df = df.set_index(["School", "Grade"])
df.head()

Name  Gender  Weight  \
School                        Grade                                       
Shanghai Jiao Tong University Freshman     Gaopeng Yang  Female    46.0   
Peking University             Freshman   Changqiang You    Male    70.0   
Shanghai Jiao Tong University Senior            Mei Sun    Male    89.0   
Fudan University              Sophomore    Xiaojuan Sun  Female    41.0   
                              Sophomore     Gaojuan You    Male    74.0   

                                        Transfer  
School                        Grade               
Shanghai Jiao Tong University Freshman         N  
Peking University             Freshman         N  
Shanghai Jiao Tong University Senior           N  
Fudan University              Sophomore        N  
                              Sophomore        N

In [170]:
df = df.sort_index()
df

Name  Gender  Weight Transfer
School              Grade                                              
Fudan University    Freshman   Changqiang Yang  Female    49.0        N
                    Freshman      Gaoqiang Qin  Female    63.0        N
                    Freshman      Gaofeng Zhao  Female    43.0        N
                    Freshman      Yanquan Wang  Female    55.0        N
                    Freshman         Feng Wang    Male    74.0        N
...                                        ...     ...     ...      ...
Tsinghua University Sophomore        Feng Qian  Female    43.0        N
                    Sophomore           Li Qin    Male    76.0        N
                    Sophomore      Yanjuan You    Male    55.0        N
                    Sophomore   Xiaoqiang Qian    Male    73.0        N
                    Sophomore      Chunpeng Lv    Male    51.0        N

[200 rows x 4 columns]

In [172]:
# 元组
df.loc[("Peking University", "Freshman")]

Name  Gender  Weight Transfer
School            Grade                                            
Peking University Freshman  Changqiang You    Male    70.0        N
                  Freshman     Gaopeng Shi  Female    48.0        N
                  Freshman      Gaoli Zhao    Male    78.0        N
                  Freshman    Xiaojuan Qin    Male    79.0        Y
                  Freshman       Qiang Han    Male    87.0        N
                  Freshman        Quan Chu  Female    43.0        N
                  Freshman    Xiaojuan Chu    Male    58.0        Y
                  Freshman   Changquan Chu  Female    45.0        N
                  Freshman    Xiaopeng Han  Female    53.0        N
                  Freshman   Changmei Feng  Female    56.0        N
                  Freshman  Changpeng Zhao    Male    83.0        N
                  Freshman     Qiang Zhang  Female    43.0        N
                  Freshman    Chunmei Wang  Female    43.0        N

In [174]:
df.loc[ [("Peking University", "Freshman"), ("Tsinghua University", "Senior")] ]

Name  Gender  Weight Transfer
School              Grade                                            
Peking University   Freshman  Changqiang You    Male    70.0        N
                    Freshman     Gaopeng Shi  Female    48.0        N
                    Freshman      Gaoli Zhao    Male    78.0        N
                    Freshman    Xiaojuan Qin    Male    79.0        Y
                    Freshman       Qiang Han    Male    87.0        N
                    Freshman        Quan Chu  Female    43.0        N
                    Freshman    Xiaojuan Chu    Male    58.0        Y
                    Freshman   Changquan Chu  Female    45.0        N
                    Freshman    Xiaopeng Han  Female    53.0        N
                    Freshman   Changmei Feng  Female    56.0        N
                    Freshman  Changpeng Zhao    Male    83.0        N
                    Freshman     Qiang Zhang  Female    43.0        N
                    Freshman    Chunmei Wang  Female    43.0        N
Tsinghua University Senior      Xiaomei Zhou  Female    57.0        N
                    Senior      Xiaofeng Sun    Male    71.0        N
                    Senior       Chunmei You    Male    69.0        N
                    Senior      Xiaopeng Chu  Female    42.0        N
                    Senior             Li Xu  Female    53.0        N
                    Senior         Juan Zhao  Female    47.0        N
                    Senior      Xiaofeng Shi  Female    55.0        N
                    Senior          Juan You  Female    40.0        N
                    Senior          Gaoli Wu  Female    47.0      NaN
                    Senior       Yanli Zhang  Female    41.0        N
                    Senior         Peng Wang    Male    73.0        N
                    Senior         Mei Zheng  Female    50.0        N
                    Senior     Xiaoqiang Qin    Male    79.0        N
                    Senior           Li Zhao  Female    50.0        N

In [179]:
# 交叉组合
# 得到所有北大和复旦的大二大三学生
df.loc[(['Peking University', 'Fudan University'], ['Sophomore', 'Junior']), :]

Name  Gender  Weight Transfer
School            Grade                                              
Peking University Sophomore      Changmei Xu  Female    43.0        N
                  Sophomore     Xiaopeng Qin    Male     NaN        N
                  Sophomore           Mei Xu  Female    39.0        N
                  Sophomore      Xiaoli Zhou  Female    55.0        N
                  Sophomore         Peng Han  Female    34.0      NaN
                  Junior             Juan Xu  Female     NaN        N
                  Junior       Changjuan You  Female    47.0        N
                  Junior            Gaoli Xu  Female    48.0        N
                  Junior        Gaoquan Zhou    Male    70.0        N
                  Junior           Qiang You  Female    56.0        N
                  Junior        Chengli Zhao    Male     NaN      NaN
                  Junior      Chengpeng Zhao  Female    44.0        N
                  Junior       Xiaofeng Zhao  Female    46.0        N
Fudan University  Sophomore     Xiaojuan Sun  Female    41.0        N
                  Sophomore      Gaojuan You    Male    74.0        N
                  Sophomore   Xiaoquan Zhang  Female    45.0        N
                  Sophomore           Mei Xu    Male    79.0        N
                  Sophomore    Chengqiang Lv  Female    53.0        N
                  Sophomore     Xiaojuan Chu    Male    68.0        N
                  Sophomore  Changqiang Qian    Male    64.0        N
                  Sophomore           Li Sun  Female    57.0        N
                  Junior           Yanli You  Female    48.0        N
                  Junior       Chunqiang Chu    Male    72.0        N
                  Junior        Changfeng Lv    Male    76.0        N
                  Junior          Yanjuan Lv  Female    49.0      NaN
                  Junior       Gaoqiang Zhou  Female    43.0        N
                  Junior       Xiaojuan Zhao  Female    49.0        N
                  Junior         Gaoquan Chu  Female    51.0        N
                  Junior         Chengli Sun    Male    62.0        N
                  Junior        Gaojuan Qian  Female    44.0        N
                  Junior       Xiaojuan Qian  Female    51.0        N
                  Junior      Chunjuan Zhang  Female    47.0        N
                  Junior        Xiaojuan Sun  Female    46.0        N

In [180]:
np.random.seed(0)
L1,L2 = ['A','B','C'],['a','b','c']
mul_index1 = pd.MultiIndex.from_product([L1,L2],names=('Upper', 'Lower'))

L3,L4 = ['D','E','F'],['d','e','f']
mul_index2 = pd.MultiIndex.from_product([L3,L4],names=('Big', 'Small'))
df_ex = pd.DataFrame(np.random.randint(-9,10,(9,9)),
                        index=mul_index1,
                        columns=mul_index2)

df_ex

Big          D        E        F      
Small        d  e  f  d  e  f  d  e  f
Upper Lower                           
A     a      3  6 -9 -6 -6 -2  0  9 -5
      b     -3  3 -8 -3 -2  5  8 -4  4
      c     -1  0  7 -4  6  6 -9  9 -6
B     a      8  5 -2 -9 -8  0 -9  1 -6
      b      2  9 -7 -9 -9 -5 -4 -3 -1
      c      8  6 -5  0  1 -8 -8 -2  0
C     a     -6 -3  2  5  9 -9  5 -6  3
      b      1  2 -5 -3 -5  6 -6  3 -5
      c     -1  5  6 -6  6  4  7  8 -4

In [181]:
pd.IndexSlice

三级索引

In [182]:
np.random.seed(0)
L1,L2,L3 = ['A','B'],['a','b'],['alpha','beta']
mul_index1 = pd.MultiIndex.from_product([L1,L2,L3],
                        names=('Upper', 'Lower','Extra'))

L4,L5,L6 = ['C','D'],['c','d'],['cat','dog']
mul_index2 = pd.MultiIndex.from_product([L4,L5,L6],
                names=('Big', 'Small', 'Other'))

df_ex = pd.DataFrame(np.random.randint(-9,10,(8,8)),
                    index=mul_index1,
                    columns=mul_index2)
df_ex

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     alpha   3   6  -9  -6  -6  -2   0   9
            beta   -5  -3   3  -8  -3  -2   5   8
      b     alpha  -4   4  -1   0   7  -4   6   6
            beta   -9   9  -6   8   5  -2  -9  -8
B     a     alpha   0  -9   1  -6   2   9  -7  -9
            beta   -9  -5  -4  -3  -1   8   6  -5
      b     alpha   0   1  -8  -8  -2   0  -6  -3
            beta    2   5   9  -9   5  -6   3   1